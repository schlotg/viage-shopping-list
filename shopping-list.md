# Viage Shopping List Tutorial

## Adding a Shopping List Component
