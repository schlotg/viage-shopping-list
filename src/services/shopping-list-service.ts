import { Service } from 'viage';

export interface Item {
  name: string;
  description: string;
  quantity: number;
  enabled: boolean;
  _id?: string; // added by the service
}

class ShoppingListServiceSingleton extends Service {

  private list: Item[] = [];
  private count = 1;

  private generateId() {
    this.count += 1;
    return this.count.toString(16);
  }

  private findIndex(id: string): number {
    let i = 0;
    const list = this.list;
    for (; i < list.length; ++i) {
      if (list[i]._id === id) { break; }
    }
    return i;
  }

  constructor() {
    super();
    if (window.localStorage) {
      const json = window.localStorage.getItem('ShoppingList');
      if (json) {
        const obj = JSON.parse(json) || {};
        this.count = obj.count || 0;
        const list = obj.list || [];
        list.forEach((e: Item) => this.addItem(e));
      }
    }
  }

  addItem(item: Item) {
    item._id = item._id || this.generateId();
    this.list.push(item);
    this.save();
  }

  removeItem(id: string) {
    const i = this.findIndex(id);
    this.list.splice(i, 1);
    this.save();
  }

  getItem(id: string): Item {
    const i = this.findIndex(id);
    return this.list[i];
  }

  forEach(cb: (item: Item, i?: number) => void) {
    this.list.forEach(cb);
  }

  save() {
    if (window.localStorage) {
      window.localStorage.setItem('ShoppingList',
        JSON.stringify({ list: this.list, count: this.count }));
    }
    this.dispatchEvent('update', this.list);
  }

  clear() {
    this.list = [];
    this.save();
  }
}

const service = new ShoppingListServiceSingleton();
export const ShoppingListService = service;
