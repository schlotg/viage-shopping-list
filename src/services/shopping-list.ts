import { Service } from 'viage';

  export class ShoppingList extends Service {
    constructor() {
      super();
    }
  }
  