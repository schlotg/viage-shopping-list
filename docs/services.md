# Viage Shopping List Tutorial

## Services
In Viage, Services are used to manage data or any other global resource. Components should only contain other components and the glu-logic to display data to the user. In contrast, Services should not contain any display logic and handle all aspects of managing the data. This includes notifying components when the data has changed. Services are singletons and do need to be instantiated, you can use them directly by importing them.

### Create a Service
Using the Viage CLI, create a new service by typing the following in a shell in your project's direrctory:

``` viage service shopping-list-service```

This will create a new file at src/services/shopping-list-service.ts. Open this file and it should look like this:

```Javascript
import { Service } from 'viage';

class ShoppingListServiceSingleton extends Service {
  constructor() {
    super();
  }
}
const service = new ShoppingListServiceSingleton ();
export const ShoppingListService = service;
```

Services extend the base class Service, and then export a single instance of the service. To use it in a component you just need to import it:

```Javascript
import { ShoppingListService } from '../service/shopping-list-service';
```

### Defining a Data interface
Let's start by defining an interface for the data. Insert the following lines just above the class declaration:

```Javascript
import { Service } from 'viage';

export interface Item {
  name: string;
  description: string;
  quantity: number;
  purchased: boolean;
  _id?: string; // added by the service
}

class ShoppingListServiceSingleton extends Service {
```

Each Shopping list item is going to have a name, description, quantity, and a purchased flag to indicate if it has been purchased or not. The _id field will be used internally by the service to uniquely track each item. We export the interface so it can be used by components interacting with the data.

### Data Retrieval and Storage
Normally a service would save and retrieve data from a server. Viage doesn't try to duplicate functionality that is already in the DOM. I reccomend using the [fetch API](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch). It is an excellant modern interface that supports promises and is available across the 4 big browsers.

Since writing server code is beyond the scope of this tutorial, we will write to the browser's local storage instead. Add the following code to the shopping-list-service:

```Javascript
class ShoppingListServiceSingleton extends Service {

  private list: Item[] = [];
  private count = 1;

  private generateId() {
    this.count += 1;
    return this.count.toString(16);
  }

  constructor() {
    super();
    if (window.localStorage) {
      const json = window.localStorage.getItem('ShoppingList');
      if (json) {
        const obj = JSON.parse(json) || {};
        this.count = obj.count || 0;
        const list = obj.list || [];
        list.forEach((e: Item) => this.addItem(e));
      }
    }
  }

  save() {
    if (window.localStorage) {
      window.localStorage.setItem('ShoppingList',
        JSON.stringify({ list: this.list, count: this.count }));
    }
    this.dispatchEvent('update', this.list);
  }
```

Our service now contains a local list of the data in a variable named *list*. There is also a local variable called *count* that the *generateId()* function uses to generate unique ids. The *contructor()* and the *save()* function contain code that stores and retrieves data from local storage.

Notice that everytime the *save()* function is called, we dispatch an event called *update*, that lets everyone that is listening know that the data has changed. The *dispatchEvent()* function is defined in the *Service* base class along with the ability to add and remove event listeners. More on that later.

### Getting and Setting Data
Next, we will add some getters and setters to make it easy to add, get, and remove data for the components. This is just standard Typescript. Add the following lines to our service:

```Javascript
  private findIndex(id: string): number {
    let i = 0;
    const list = this.list;
    for (; i < list.length; ++i) {
      if (list[i]._id === id) { break; }
    }
    return i;
  }

  addItem(item: Item) {
    item._id = item._id || this.generateId();
    this.list.push(item);
    this.save();
  }

  removeItem(id: string) {
    const i = this.findIndex(id);
    this.list.splice(i, 1);
    this.save();
  }

  getItem(id: string): Item {
    const i = this.findIndex(id);
    return this.list[i];
  }

  forEach(cb: (item: Item, i?: number) => void) {
    this.list.forEach(cb);
  }

  clear() {
    this.list = [];
    this.save();
  }
```

The function *findIndex* is just a helper function to find an element by id so that we can remove and get items. There is also a *forEach()* function that makes it easy to iterate across all of the items in our list.

Whew! That is the biggest file in our project but it actually does a lot and is great foundation for the rest of the app. All we need to do now is bolt up some components to display the data.

The final file should look like this:

```Javascript
import { Service } from 'viage';

export interface Item {
  name: string;
  description: string;
  quantity: number;
  purchased: boolean;
  _id?: string; // added by the service
}

class ShoppingListServiceSingleton extends Service {

  private list: Item[] = [];
  private count = 1;

  private generateId() {
    this.count += 1;
    return this.count.toString(16);
  }

  constructor() {
    super();
    if (window.localStorage) {
      const json = window.localStorage.getItem('ShoppingList');
      if (json) {
        const obj = JSON.parse(json) || {};
        this.count = obj.count || 0;
        const list = obj.list || [];
        list.forEach((e: Item) => this.addItem(e));
      }
    }
  }

  save() {
    if (window.localStorage) {
      window.localStorage.setItem('ShoppingList',
        JSON.stringify({ list: this.list, count: this.count }));
    }
    this.dispatchEvent('update', this.list);
  }

  private findIndex(id: string): number {
    let i = 0;
    const list = this.list;
    for (; i < list.length; ++i) {
      if (list[i]._id === id) { break; }
    }
    return i;
  }

  addItem(item: Item) {
    item._id = item._id || this.generateId();
    this.list.push(item);
    this.save();
  }

  removeItem(id: string) {
    const i = this.findIndex(id);
    this.list.splice(i, 1);
    this.save();
  }

  getItem(id: string): Item {
    const i = this.findIndex(id);
    return this.list[i];
  }

  forEach(cb: (item: Item, i?: number) => void) {
    this.list.forEach(cb);
  }

  clear() {
    this.list = [];
    this.save();
  }
}

const service = new ShoppingListServiceSingleton();
export const ShoppingListService = service;

```

### [Next (Adding a Shopping List Element Component)](shopping-list-element.md)