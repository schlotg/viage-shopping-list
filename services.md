# Viage Shopping List Tutorial

## Services
In Viage Services are used to manage data or any other global resource. Components should only contain other components and the glu-logic to display data to the user. In contrast Services should not contain and display logic and handle all aspects of managing the data that is being displayed including notifying components when data has changed. Services are singletons and do need to be instantiated, you cna use them directly by importing them.

### Create a Service
Using the Viage CLI, create a new service by typing the following in a shell in your project's direrctory:

``` viage service shopping-list-service```

This will create a new file at src/services/shopping-list-service.ts. Open this file and it should look like this:

```Javascript
import { Service } from 'viage';

class ShoppingListServiceSingleton extends Service {
  constructor() {
    super();
  }
}
const service = new ShoppingListServiceSingleton ();
export const ShoppingListService = service;
```

Services extend the base class Service, and then exports a single instance of the service. To use it in a component you just need to import it:

```Javascript
import { ShoppingListService } from '../service/shopping-list-service';
```

### Defining a Data interface
Let's start by defining an interface for the data. Insert the following lines just above the class declaration:

```Javascript
import { Service } from 'viage';

export interface Item {
  name: string;
  description: string;
  quantity: number;
  purchased: boolean;
  _id?: string; // added by the service
}

class ShoppingListServiceSingleton extends Service {
```

Each Shopping list item is going to have a name, description, quantity, and a purchased flag to indicated if purchased or not. The _id field will be used internally by the service to uniquely track each item. We export the interface so it can be used by components interacting with the data.



