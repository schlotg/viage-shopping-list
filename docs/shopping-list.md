# Viage Shopping List Tutorial

## Adding a Shopping List Component
Now we need a component to display all the shopping list elements. Start by using the CLI to create a *shopping-list component*.

``` viage component shopping-list ```

This will create the the file src/components/shopping-list.ts

## Modify the Code
Now modify the code in the new component so lit looks like the following:

```Javascript
import { Component } from 'viage';
import { ShoppingListService } from '../services/shopping-list-service';
import { ShoppingListElement } from './shopping-list-element';
import { States } from './app';

export class ShoppingList extends Component {

  constructor(){
    super('shopping-list');
  }

  init() {
    this.setHTML(`
      <button attach="add" style="color:#00c700">Add</button>
      <button attach="clear" style="color:#ff6e6e">Clear</button>
      <div attach="list" style="margin-top:20px; width:630px; padding:5px; background-color: #eeeeee"></div>
    `);
    this.updateList();
    this.addServiceListener(ShoppingListService, 'update', () => this.updateList());
    this.attachments.clear.addEventListener('click', () => ShoppingListService.clear());
    this.attachments.add.addEventListener('click', () => {
      const addUrl = this.router.createUrl<void>(States.ADD);
      this.router.go(addUrl);
    });
    return this;
  }

  updateList(){
    this.clearComponents();
    this.attachments.list.innerHTML = '';
    ShoppingListService.forEach(e => {
      this.createComponent(ShoppingListElement).init(e)
        .attach(this.attachments.list);
    });
  }
}
```

### List HTML
There are several things worth noting here. This component creates a **Add** button, **Clear** button, and a div that acts as a container to render the *ShoppingListElements* into. Just like the previous step we add click handlers to our buttons which will clear the list and tell the router to go to the Add state.

### Interacting with the Service
This component installs a update event listener on the *ShoppingListService*. If you recall, this gets fired everytime the data changes in the service. You'll note that we use the Viage Component API  *addServiceListener()* to add the event listener. This is because this function has a nice cleanup mechanism that automatically removes the listener when this component gets destroyed. It is not neccesary to do this with event listeners in the owning component's HTML. This due to the fact that the Component will destroy its HTML elements when it goes away, which in turn destroys the event listeners.

### Drawing the List
The *updateList()* function first calls *clearComponents()* which clears out all the child components that this component owns. Next the containers children are cleared out by assigning the container's inner HTML to a empty string.

Next we call the *ShoppingListService.forEach()* to get each element from the service. Then we call the *createComponent()* function to create the *ShoppingListElement* and init it with the item's data. The *createComponent()* function is a Viage API that creates the component and then adds it to *this.components[]* . When the component gets destroyed, it will automatically destroy all of its child components as well.

You will note that the *attach()* function is called on each newly created Component. The container element is passed in so that each child component will render into the container.

### Drawing the shopping-list Component
If you have been paying attention, you will have noticed that the *shopping-list-elements* render into the shopping-list component, but so far the shopping-list component is not attached to the DOM. In the next section we will take care of that by using the router.

### [Next (The Router)](routes.md)
