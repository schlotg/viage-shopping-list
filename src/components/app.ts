import { ShoppingList } from './shopping-list';
import { ShoppingListAdd } from './shopping-list-add';
import { createRouter, Component } from 'viage';

export enum States {
  HOME = 'home',
  ADD = 'add',
  EDIT = 'edit'
};

export class App extends Component {

  private title = 'Shopping List';

  constructor() {
    super('app');
  }
  init() {
    document.querySelector('title').textContent = this.title;

    this.attach('page', true);
    this.setHTML(`
      <h1 style="text-align: center">${this.title}</h1>
      <div style="width: 100%; min-width: 654px" attach="portal"></div>
    `);

    const router = createRouter('main', this.attachments.portal, 'HASH');
    router.addStates([
      { name: States.HOME, component: ShoppingList,  type: 'DEFAULT' },
      { name: States.ADD, component: ShoppingListAdd,  type: 'NORMAL' },
      { name: States.EDIT, component: ShoppingListAdd,  type: 'NORMAL' },
    ]);
    // start off by going to the state the page was loaded on
    router.start();
    return this;
  }
}
