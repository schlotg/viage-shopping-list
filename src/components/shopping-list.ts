import { Component } from 'viage';
import { shoppingListService, Item } from '../services/shopping-list-service';
import { ShoppingListElement } from './shopping-list-element';
import { States } from './app';

export class ShoppingList extends Component {

  constructor(){
    super('shopping-list');
  }

  init() {
    this.setHTML(`
      <div class="logo-img"></div>
      <button attach="add" class="button green">Add</button>
      <button attach="clear" class="button red">Clear</button>
      <div attach="list" class="shopping-list"></div>
    `);
    this.updateList();
    this.addServiceListener<Item[]>(shoppingListService, 'update', (list: Item[]) => this.updateList());
    this.a.clear.addEventListener('click', () => shoppingListService.clear());
    this.a.add.addEventListener('click', () => {
      const addUrl = this.router.createUrl<void>(States.ADD);
      this.router.go(addUrl);
    });
    return this;
  }

  updateList(){
    this.clearComponents();
    this.a.list.innerHTML = '';
    shoppingListService.forEach(e => {
      this.createComponent(ShoppingListElement).init(e)
        .attach(this.a.list);
    });
  }
}
