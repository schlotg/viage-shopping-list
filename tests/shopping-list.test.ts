import { ShoppingListService, Item } from '../src/services/shopping-list-service';
import { ShoppingList } from '../src/components/shopping-list';
import { createRouter } from 'viage';
import { item1, items, triggerClickEvent } from './common';

const consoleWarnMock = jest.spyOn(console, 'warn');
consoleWarnMock.mockImplementation(() => {});
const consoleErrorMock = jest.spyOn(console, 'error');
consoleErrorMock.mockImplementation(() => {});

test('ShoppingList should initialize correctly', () => {
  const mock = jest.spyOn(ShoppingListService, 'forEach');
  mock.mockImplementation((cb) => items.forEach(cb));
  const shoppingList = new ShoppingList();
  shoppingList.init();
  expect(shoppingList.e.innerHTML).toMatchSnapshot();
  const components = [];
  shoppingList.forEachComponents((component) => components.push(component));
  expect(components.length).toBe(2);
  mock.mockRestore();
});

test('Clear handler should call service.clear()', () => {
  const mock = jest.spyOn(ShoppingListService, 'clear');
  const shoppingList = new ShoppingList();
  shoppingList.init();
  const clear = shoppingList.e.querySelector('[attach=clear]') as HTMLElement;
  triggerClickEvent(clear);
  expect(mock).toBeCalled();
  mock.mockRestore();
});

test('Add handler should call router.go()', () => {
  const router = createRouter('test', document.createElement('div'), 'LOCATION' );
  const shoppingList = new ShoppingList();
  shoppingList.init();
  const add = shoppingList.e.querySelector('[attach=add]') as HTMLElement;
  const mock = jest.spyOn(router, 'go');
  const mock2 = jest.spyOn(router, 'createUrl');
  shoppingList.setRouter(router);
  triggerClickEvent(add);
  expect(mock).toBeCalled();
  mock.mockRestore();
  mock2.mockRestore();
});

test('Adding a item to the ShoppingListService should trigger an updateList() call', () => {
  const router = createRouter('test', document.createElement('div'), 'LOCATION' );
  const shoppingList = new ShoppingList();
  shoppingList.init();
  const mock = jest.spyOn(shoppingList, 'updateList');
  shoppingList.setRouter(router);
  ShoppingListService.addItem(item1);
  expect(mock).toBeCalled();
  mock.mockRestore();
});


