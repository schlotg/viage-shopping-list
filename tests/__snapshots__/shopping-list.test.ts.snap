// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingList should initialize correctly 1`] = `
"
      <button attach=\\"add\\" style=\\"color:#00c700\\">Add</button>
      <button attach=\\"clear\\" style=\\"color:#ff6e6e\\">Clear</button>
      <div attach=\\"list\\" style=\\"margin-top:20px; width:630px; padding:5px; background-color: #eeeeee\\"><shopping-list-element>
      <div style=\\"width: 100%; margin-top: 5px; margin-bottom: 5px;\\">
        <input attach=\\"enabled\\" type=\\"checkbox\\">
        <span style=\\"width: 30px; display: inline-block\\">1</span>
        <span style=\\"width: 100px; display: inline-block\\">Dog Food</span>
        <span style=\\"width: 300px; display: inline-block\\">30 lb bag</span>
        <button attach=\\"delete\\" style=\\"padding: 1px;\\">Delete</button>
        <button attach=\\"edit\\" style=\\"padding: 1px;\\">Edit</button>
      </div>
    </shopping-list-element><shopping-list-element>
      <div style=\\"width: 100%; margin-top: 5px; margin-bottom: 5px;\\">
        <input attach=\\"enabled\\" type=\\"checkbox\\">
        <span style=\\"width: 30px; display: inline-block\\">12</span>
        <span style=\\"width: 100px; display: inline-block\\">Foo Bars</span>
        <span style=\\"width: 300px; display: inline-block\\">1 Box</span>
        <button attach=\\"delete\\" style=\\"padding: 1px;\\">Delete</button>
        <button attach=\\"edit\\" style=\\"padding: 1px;\\">Edit</button>
      </div>
    </shopping-list-element></div>
    "
`;
